app/services/preserved_object_handler.rb

```diff
@@ -33,56 +38,99 @@ class PreservedObjectHandler
-                             size: incoming_size,
-                             preservation_policy: pp_default)
+      po = PreservedObject.create!(druid: druid,
+                                   current_version: incoming_version,
+                                   size: incoming_size,
+                                   preservation_policy: pp_default)
+      status = Status.find_by(status_text: "ok") # TODO: Replace the status_text to get a default status_text
+      PreservationCopy.create(preserved_object: po,
+                              current_version: incoming_version,
+                              last_audited: nil,
+                              endpoint: endpoint,
+                              status: status,
+                              last_checked_on_storage: nil)
       results << result_hash(CREATED_NEW_OBJECT)
     end
     results.flatten!
     log_results(results)
     results
   end
 
+  def update
+    results = []
+    if invalid?
+      results << result_hash(INVALID_ARGUMENTS, errors.full_messages)
+    elsif !PreservedObject.exists?(druid: druid)
+      results << result_hash(OBJECT_DOES_NOT_EXIST, 'PreservedObject')
+    else
+      Rails.logger.debug "update #{druid} called and object exists"
+      begin
+        po_db_object = PreservedObject.find_by(druid: druid)
+        results << update_per_version_comparison_po(po_db_object)
+        pc_db_object = PreservationCopy.find_by(preserved_object: po_db_object, endpoint: endpoint)
+        results << update_per_version_comparison_pc(pc_db_object)
+      rescue ActiveRecord::ActiveRecordError => e
+        results << result_hash(DB_UPDATE_FAILED, "#{e.inspect} #{e.message} #{e.backtrace.inspect}")
+      end
+    end
+    results.flatten!
+    log_results(results)
+    results
+  end
+
   private
 
   # expects @incoming_version to be numeric
   # TODO: update existence check timestamps/status per each flavor of comparison?
-  def update_per_version_comparison(db_object)
+  def update_per_version_comparison_po(db_object)
     version_comparison = db_object.current_version <=> incoming_version
     results = []
     if version_comparison.zero?
-      results << result_hash(VERSION_MATCHES)
+      results << result_hash(VERSION_MATCHES, db_object.class.name)
     elsif version_comparison == 1
       # TODO: needs manual intervention until automatic recovery services implemented
-      results << result_hash(ARG_VERSION_LESS_THAN_DB_OBJECT)
+      results << result_hash(ARG_VERSION_LESS_THAN_DB_OBJECT, db_object.class.name)
     elsif version_comparison == -1
       db_object.current_version = incoming_version
       db_object.size = incoming_size if incoming_size
-      results << result_hash(ARG_VERSION_GREATER_THAN_DB_OBJECT)
+      results << result_hash(ARG_VERSION_GREATER_THAN_DB_OBJECT, db_object.class.name)
     end
+    update_db_object(db_object, results)
+    results.flatten
+  end
 
+  def update_per_version_comparison_pc(db_object)
```

Comment:
am I missing something obvious?  it seems like update_per_version_comparison_xx take the same argument, have the same comparison, do the same thing, and call the same method ... so why do we have separate methods?  line 115 could be in an "if is_instance_of?" clause.   

If there's a reason to keep them distinct, would it make sense to distinguish the param name (preserved_object vs preservation_copy) ?

Reply:
No reason. We just wanted to get the code working and refactor later - i'll update this :)